CI: agent tests can hit tool_timeout and surface "TimeoutError: " as a result, breaking string assertions

Every tool call in the `SerenaAgent` integration tests goes through `Tool.apply_ex` → `TaskExecutor` → `future.result(timeout=tool_timeout)`, with the global default of 240s (`DEFAULT_TOOL_TIMEOUT`, `src/serena/config/serena_config.py`). On timeout, `apply_ex` catches the exception and, on the default `catch_exceptions=True` path, returns the literal string `"TimeoutError: "` as the tool result. Assertions on result content then fail with a confusing message — and negative assertions (`assert msg not in result`) pass vacuously. Both happened in the run below.

## Observed (2026-07-10)

Run https://github.com/oraios/serena/actions/runs/29064865197, `catch-all (ubuntu-latest)` — 933 passed, 1 failed:

```
FAILED test/serena/test_serena_agent.py::TestPromptProvision::test_dynamically_activated_mode_is_provided_once_per_session[python]
AssertionError: Expected new mode message 'The onboarding process is not applied.' to be included in new session, but it was not found in result:
  TimeoutError:
```

From the job log:

- The `ActivateProjectTool` call (python → java test repo) completed in time. The two **post-activation** `InitialInstructionsTool` calls each hit the 240s limit — log lines at 02:41:44 and 02:45:44, exactly 240s apart — apparently queued behind the Java LS startup on the executor's single worker thread.
- This was under two minutes after job start (02:35:56 → 02:37:43) on a fresh runner, not under end-of-job load.
- The intermediate negative assertion ("mode message must NOT appear again in the same session") passed vacuously against `"TimeoutError: "`; only the final positive assertion surfaced anything.

Second incident on the same run: `native (ubuntu-latest)` ran 02:36:34 → 08:36:50 and was killed at GitHub's default 6-hour job limit — the `cpu` matrix job sets no `timeout-minutes`.

## Scope

Not isolated to this test. Everything using the `serena_agent` fixture shares the path: ~23 methods in `TestSerenaAgent` / `TestPromptProvision` expanding to dozens of parametrized items, all in the catch-all batch. `pytest-timeout` is a dev dependency but not wired into `[tool.pytest.ini_options]`, and nothing under `test/` overrides `tool_timeout`. The failing test is simply the heaviest member (cross-project activation plus follow-up calls), so it surfaced first.

## Proposed fixes

Each independent, each its own PR:

1. **`timeout-minutes` on the `cpu` matrix job** — backstop for the 6-hour-runaway class.
   PR follows immediately; value derived from the slowest healthy job across the matrix plus margin.
2. **Self-explanatory timeout message in `apply_ex`** — tool name plus configured timeout instead of the bare `"TimeoutError: "`.
   Continues the direction of #1582 (failures legible to the client); production MCP clients currently receive the same bare string.
3. **`catch_exceptions=False` in the test helper `_call_tool`** — a timeout raises instead of poisoning the result string.
   Fixes both observed modes (confusing failure, vacuous pass) at once. Minimal flip only; call sites that intentionally assert on error strings keep current behavior.
4. **Per-class `pytest-timeout` markers** on the agent test classes, with generous values (precedent: the Erlang tests already carry `@pytest.mark.timeout(60)`).
   Deliberately not a tight global default: signal-based timeouts don't fire off the main thread, thread-based teardown interacts badly with LS subprocesses and the executor's background threads, and a marginal global value would just be a new flake source.

## Open questions

- **Fixture cost and batch placement.** `serena_config` registers 17 languages regardless of a test's parametrization, and `serena_agent` is function-scoped (full agent + LS start/stop per item), all inside the longest batch. Both touch the CI batch design (related: #1663), so raising here rather than proposing changes.
- **Root cause.** Should post-activation tool calls queue behind LS init on the executor at all, or do they need a separate budget (related: #1052)? The activation call itself succeeded; the follow-ups starved. Worth its own investigation — happy to dig if wanted.


Preconditions:

- [x] I have made sure it's an actual issue, not a question (use [GitHub Discussions](https://github.com/oraios/serena/discussions) instead).
- [x] I have consulted the [user guide](https://oraios.github.io/serena/02-usage/) and verified that the issue cannot be resolved by adjusting configuration/following recommended workflows.
- [x] I have looked for similar issues and discussions, including closed ones.

Issue details:

- [x] I have provided a meaningful title and description.
- [x] I have explained how the issue arose and, where possible, added instructions on how to reproduce it.
- [ ] I have added details on my setup: Serena version, MCP client, OS, the programming language(s), relevant configuration adjustments and project specifics.
- [x] If the issue relates to an application of Serena to an open-source project, I have added the link.

--- a/src/serena/config/serena_config.py
+++ b/src/serena/config/serena_config.py
@@ -873,6 +873,9 @@ class SerenaConfig(SharedConfig, ModeSelectionDefinitionWithBaseModes):
     JetBrains IDE launch command, which can be used to auto-start an IDE instance on demand.
     """
     tool_timeout: float = DEFAULT_TOOL_TIMEOUT
+    """
+    timeout for tool calls in seconds; if a tool takes longer than this, it is aborted and an error is returned.
+    """
 
     token_count_estimator: str = RegisteredTokenCountEstimator.CHAR_COUNT.name
     """Only relevant if `record_tool_usage` is True; the name of the token count estimator to use for tool usage statistics.

--- a/test/serena/test_serena_agent.py
+++ b/test/serena/test_serena_agent.py
@@ -764,7 +764,7 @@ SAFE_DELETE_SUCCEEDS_CASES = [
 
 @pytest.fixture
 def serena_config():
-    config = SerenaConfig(log_level=logging.ERROR).with_headless_mode_overrides()
+    config = SerenaConfig(log_level=logging.ERROR, tool_timeout=600).with_headless_mode_overrides()
 
     # Create test projects for all supported languages
     test_projects = []
@@ -1047,7 +1047,7 @@ class TestSerenaAgent:
         agent = serena_agent
 
         find_symbol_tool = agent.get_tool(FindSymbolTool)
-        result = find_symbol_tool.apply_ex(
+        result = find_symbol_tool.apply(
             name_path_pattern=case.name_path,
             depth=0,
             relative_path=None,
@@ -1070,7 +1070,7 @@ class TestSerenaAgent:
         agent = serena_agent
 
         find_symbol_tool = agent.get_tool(FindSymbolTool)
-        result = find_symbol_tool.apply_ex(
+        result = find_symbol_tool.apply(
             name_path_pattern=case.name_path,
             depth=0,
             substring_matching=True,
@@ -1088,7 +1088,7 @@ class TestSerenaAgent:
         agent = serena_agent
 
         find_symbol_tool = agent.get_tool(FindSymbolTool)
-        result = find_symbol_tool.apply_ex(
+        result = find_symbol_tool.apply(
             name_path_pattern=case.name_path,
             depth=0,
             substring_matching=False,
@@ -1342,7 +1342,7 @@ class TestPromptProvision:
 
     @classmethod
     def _call_tool(cls, agent: SerenaAgent, tool_class: type[Tool], session_id: str = "global", **kwargs) -> str:
-        result = agent.get_tool(tool_class).apply_ex(mcp_ctx=cls.MockContext(session_id), **kwargs)
+        result = agent.get_tool(tool_class).apply_ex(mcp_ctx=cls.MockContext(session_id), catch_exceptions=False, **kwargs)
         return result
 
     @staticmethod

